### ISSUE TITLE
Support more than 256 leds

### ISSUE THREAD

**Issue by mateuszdrwal** [qmk/qmk_firmware#24675]
I made a split keyboard with 4 leds under each key (10 for the 2 2U keys) and the results are [pretty spectacular](https://github.com/user-attachments/assets/8e028842-690f-4ba8-81c3-1d9bc073de4a). This sums up to 162 leds per half, or 324 leds on both. QMK currently uses a `uint_8` in most places to iterate over leds, so this required some patches which I want to polish up and contribute back.

I'm not intimately familiar with the codebase or nuances of QMK, so it would be great with some guidance on the best way to implement it for this project. I assume RAM is enough of a concern for the AVR platform that replacing all relevant `uint_8` with `uint_16` would be problematic? Would a `typedef uintX_t num_leds_t` then be best setting the type to uint16_t if `RGB_MATRIX_LED_COUNT > 255`, and using that type everywhere instead? Anything else I should know before putting time into implementing this properly?

**Comment by drashna** [comment]
In general, yeah, AVR ram limitations are usually an issue.

I'm not sure of the best solution, but a typedef sounds like probably the best option here.  There is a similar thing for the layer state bitmask, for instance.

**Comment by getreuer** [comment]
That board is gorgeous!

Maybe the LED index type can be defined conditionally based on `RGB_MATRIX_LED_COUNT`? It would be nice if it works automatically without needing an additional configuration. Perhaps:

```c
#if (RGB_MATRIX_LED_COUNT) <= 255
typedef uint8_t led_index_t;  // Continue using uint8_t by default.
#else
typedef uint_fast16_t led_index_t;  // If there are more than 255 LEDs.
#endif

bool SOME_EFFECT(effect_params_t* params) {
    LED_MATRIX_USE_LIMITS(led_min, led_max);
    for (led_index_t i = led_min; i < led_max; i++) {
       ...
```

**Comment by mateuszdrwal** [comment]
Yep, that's what I'm thinking. `led_index_t` is probably a better type name than what I came up with.

**Comment by christrotter** [comment]
I ran into this on my last board, but dang, your implementation of 'maxing out led count' is far more wondrous.

**Comment by lazizbekravshanov** [comment]
Happy to help push this forward. One thing that doesn't seem to be in the thread yet:

Widening the loop counters / adding a `led_index_t` typedef isn't sufficient on its own, because the index space also carries a sentinel — [`#define NO_LED 255`](https://github.com/qmk/qmk_firmware/blob/2dc5e397985102f4aa18b7e9cc81ee5d06afba74/quantum/rgb_matrix/rgb_matrix_types.h#L70-L76), with `matrix_co` declared as `uint8_t[MATRIX_ROWS][MATRIX_COLS]`. A board with more than 254 real LEDs would collide with the sentinel value even after the iterators are widened, so `matrix_co` and `NO_LED` need to widen together. The same applies in the mono `led_matrix` subsystem (`quantum/led_matrix/led_matrix_types.h`), not just `rgb_matrix`.

Proposed scope for a PR:
- Introduce a conditional `led_index_t` — stay `uint8_t` by default to protect AVR RAM, widen only when the LED count exceeds 255.
- Update `rgb_matrix_limits_t`, the `RGB_MATRIX_USE_LIMITS` locals, `matrix_co`, `NO_LED`, and the effect loop counters in lockstep.

Does that line up with the direction you and @getreuer were leaning toward?

**Comment by drashna** [comment]
In that direction, yeah.  However, even with all of those changes, when testing on actual hardware, it won't load if there are more than 255 leds.   There is something lower level that is causing issues too. 

For reference, I had this: https://github.com/qmk/qmk_firmware/commit/95e9c112c2c14f32634e3e7ec3cce6b5d6806aed

**Comment by drashna** [comment]
I think I may have this working.   @mateuszdrwal @lazizbekravshanov would you be willing to give this commit a spin?
https://github.com/drashna/qmk_firmware/commit/4a2a73488c2c103936b5a13f8c555bc6e52ed1df
